week3/homework/src/utilities.js

```diff
@@ -0,0 +1,108 @@
+'use strict';
+const fs = require('fs');
+const Joi = require('joi');
+const toDoFile = 'todolist.json';
+
+const readToDoList = () => {
+  return new Promise((resolve, reject) => {
+    if (!fs.existsSync(toDoFile)) {
+      fs.appendFile(toDoFile, '[]');
+    };
+    fs.readFile(toDoFile, (err, data) => {
+      if (err) { return reject(err); };
+      return resolve(data);
+    });
+  });
+};
+function writeData(file, data) {
+  fs.watchFile(file, JSON.stringify(data), (error) => {
+    if (error) { console.log(error); }
+  });
+};
+const validateTask = (task) => {
+  const schema = {
+    id:Joi.number().required(),
+    description : Joi.string().required(),
+    done:Joi.boolean().required()
+  };
+  return Joi.validate(task, schema);
+};
+const appendToDo = async(item, res) => {
+  const data = await readToDoList();
+  const toDoList = JSON.parse(data);
+  const itemObj = {
+    index:toDoList.length + 1,
+    description : item
+  };
+  toDoList.push(itemObj);
+  writeData(toDoFile, toDoList);
+  res.status(201);
+  res.json(itemObj);
+};
+const removeToDo = async(id, res) => {
+  const todoList = await readToDoList();
+  const removedTodo = todoList.find(task => task.id === id);
+  const index = todoList.indexOf(removedTodo);
```

Comment:
If you just use the index of an array (like in `todoList[`*0*`]`) you don't need the `id` in your object. 
You've now introduced a bug:
If I have a list of todo's:
```
[
    {
        index: 1,
        descritpion: "some description",
        done: false
    },
    {
        index: 2,
        descritpion: "some other description",
        done: false
    },
    {
        index: 3,
        descritpion: "some totally new description",
        done: false
    }
]
```
And I remove the second by making a `DELETE` request to `/todos/2`, it will look for `todoList[1]`, and remove that one. My list is now:
```
[
    {
        index: 1,
        descritpion: "some description",
        done: false
    },
    {
        index: 3,
        descritpion: "some totally new description",
        done: false
    }
]
```
When I now add a new item witch `appendToDo()`on line 34, it uses `toDoList.length + 1`, which is 3 now, os my list will now be:
```
[
    {
        index: 1,
        descritpion: "some description",
        done: false
    },
    {
        index: 3,
        descritpion: "some totally new description",
        done: false
    },
    {
        index: 3,
        descritpion: "something added after deleting an item",
        done: false
    }
]
```
You have two times `index: 3` now...
If you remove the index from the objects and just only use the `array` index, you will be fine!